src/Orleans/Messaging/GatewayConnection.cs

```diff
@@ -176,6 +176,17 @@ public void Connect()
             }
         }
 
+        protected override void DrainQueueOnStop(IEnumerable<Message> requests)
```

Comment:
Hmm...., instead of doing the protected msg hack, why not just add a method on the AsynchQueueAgent: DrainQueue and call it from within GatewayConnection.Stop, before calling base.Stop.